### PR TITLE
Add optional optimizer to remove leftover Q->DQ pairs after all other QDQ processing has completed

### PR DESCRIPTION
**Pull request by skottmckay**
**Description**: 
Add an optimizer that can remove leftover Q->DQ pairs. Depending on the model this may help with performance and/or improve accuracy.

**Motivation and Context**
Give user the option to avoid a round-trip between float -> 8-bit -> float.

### CODE REVIEW

**Comment by yufenglee** [comment]
There is another case: DequantizeLinear of Weight. It is excluded from const folding optimizer to allow the following QDQ fusion. If QDQ fusion doesn't happen, it'd better to const fold it again.We need to register const folding optimizer in level 2.

File: onnxruntime/core/optimizer/qdq_transformer/qdq_final_cleanup.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    The impact on performance and accuracy of removing the pair will depend on the model.
+
+    If it was quantized with Quantization Aware Training (QAT) it may be better to pay the performance cost of keeping
+    the pair as the model was trained with a round-trip from float -> 8-bit -> float.
+
+    If the model was quantized with Post Training Quantization (PTQ) it is most likely better to remove the pair as the
+    loss of precision from the round trip of float -> 8-bit -> float was not present when the model was trained.
+
+    As we have no knowledge of how the model was quantized we require the user to specify an option to enable this
+    transformer.
```
Comment (yufenglee) [inline onnxruntime/core/optimizer/qdq_transformer/qdq_final_cleanup.h:30]: we can add a metadata in model that is quantized by ORT quantization tool, but for quant model from tf and pytorch, we have no way to differentiate.

File: onnxruntime/core/optimizer/qdq_transformer/qdq_final_cleanup.cc [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    auto* node_ptr = graph.GetNode(node_index);
+    if (node_ptr == nullptr)
+      continue;  // node was removed as part of an earlier optimization
+
+    Node& q_node = *node_ptr;
+    ORT_RETURN_IF_ERROR(Recurse(q_node, modified, graph_level, logger));
+
+    if (!QDQ::MatchQNode(q_node) ||
+        !graph_utils::IsSupportedProvider(q_node, GetCompatibleExecutionProviders()) ||
+        !optimizer_utils::CheckOutputEdges(graph, q_node, 1)) {
```
Comment (yufenglee) [inline onnxruntime/core/optimizer/qdq_transformer/qdq_final_cleanup.cc:28]: >!optimizer_utils::CheckOutputEdges(graph, q_node, 1)) { [](http://example.com/codeflow?start=8&length=55)

do we need to cover the corner case that the output of Q is graph output?

File: include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h [hunk 3]
```diff
@@ -47,6 +47,15 @@ static const char* const kOrtSessionOptionsConfigSetDenormalAsZero = "session.se
...
 static const char* const kOrtSessionOptionsDisableQuantQDQ = "session.disable_quant_qdq";
 
+// If set to "1", enables the removal of QuantizeLinear/DequantizeLinear node pairs once all QDQ handling has been
+// completed. e.g. If after all QDQ handling has completed and we have -> FloatOp -> Q -> DQ -> FloatOp -> the
+// Q -> DQ could potentially be removed. This will provide a performance benefit by avoiding going from float to
+// 8-bit and back to float, but could impact accuracy. The impact on accuracy will be model specific and depend on
+// other factors like whether the model was created using Quantization Aware Training or Post Training Quantization.
+// As such, it's best to test to determine if enabling this works well for your scenario.
+// The default value is "0"
+static const char* const kOrtSessionOptionsEnableQuantQDQCleanup = "session.enable_quant_qdq_cleanup";
```
Comment (gwang-msft) [inline include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h:57]: Can this be enabled by default and can be disabled on demand (in case of precision loss)? Otherwise, nobody will use this.

File: onnxruntime/core/optimizer/qdq_transformer/qdq_final_cleanup.cc (hunk 2 above)
Comment (skottmckay) [reply onnxruntime/core/optimizer/qdq_transformer/qdq_final_cleanup.cc:28]: CheckOutputEdges does that I believe. One edge, no graph output for that to return true.

File: include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h (hunk 3 above)
Comment (skottmckay) [reply include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h:57]: I will add it to the mobile step-by-step docs. I don't think we can enable by default as it _could_ cause loss of accuracy for a QAT model.

**Comment by skottmckay** [comment]
> There is another case: DequantizeLinear of Weight. It is excluded from const folding optimizer to allow the following QDQ fusion. If QDQ fusion doesn't happen, it'd better to const fold it again.We need to register const folding optimizer in level 2.

I will add a task to consider this. We may need a minimal version of constant folding without dependencies on shape inferencing for that to be viable in a minimal build. Would also be good to get an idea of which operators are remaining and whether they have weights involved or not. e.g. the Softmax that is left over in mobilenet v2 doesn't involve weights.

**Review by yufenglee — DISMISSED** [review]

File: onnxruntime/test/optimizer/qdq_transformer_test.cc [hunk 4]
```diff
@@ -2140,5 +2142,113 @@ TEST(QDQTransformerTests, QDQ_Selector_Test) {
...
+                      TransformerLevel::Level2,
+                      12 /*opset_version*/,
+                      0.025f /*per_sample_tolerance*/,
+                      0.01f /*relative_per_sample_tolerance*/,
+                      std::make_unique<QDQFinalCleanupTransformer>(),
+                      &func);
+  };
+
+  test_case({{1, 2, 4}, {1, 3, 4}});
+  // test_case({{1, 2, 4}, {1, 3, 4}}, true);         // block removal of first dq
```
Comment (edgchen1) [inline onnxruntime/test/optimizer/qdq_transformer_test.cc:2211]: enable these tests too?

Comment (skottmckay) [reply onnxruntime/test/optimizer/qdq_transformer_test.cc]: Done. Was debugging a change and forgot to re-enable them.

**Review by edgchen1 — APPROVED** [review]
